### ISSUE TITLE
How can i use the kafka adapter?

### ISSUE THREAD

**Issue by David-Lio** [wesone/blackrik#3]
I found in the documentation that i need to provide in Blackrik config the brokers for the kafka event bus adapter but i cant find how can i use that adapter.

**Comment by wesone** [comment]
A kafka broker ist just a kafka server. So if you host a kafka server somewhere, you need to add its host (or IP) and port in the `brokers` array.

For example if the host `kafka.myproject.com` points to your kafka server, which is running on port `9092`, your config would contain something like:
```javascript
eventBusAdapter: {
    module: Blackrik.ADAPTERS.EVENTBUS.Kafka
    args: {
        brokers: [
            'kafka.myproject.com:9092'
        ]
    }
}
```
if you want to use docker to run your kafka server(s), this link might help: https://kafka.js.org/docs/running-kafka-in-development (which is basically what we have in our [docker-compose.yml](https://github.com/wesone/blackrik/blob/master/docker-compose.yml)).

The adapter `Blackrik.ADAPTERS.EVENTBUS.Kafka` uses [KafkaJS](https://kafka.js.org/docs/getting-started) and you should be able to use every [KafkaJS option](https://kafka.js.org/docs/configuration) inside the `args` object. For example:
```javascript
eventBusAdapter: {
    module: Blackrik.ADAPTERS.EVENTBUS.Kafka
    args: {
        brokers: [
            'kafka.myproject.com:9092'
        ],
        authenticationTimeout: 1000,
        reauthenticationThreshold: 10000,
        ssl: true,
        sasl: {
            mechanism: 'plain',
            username: 'my-username',
            password: 'my-password'
        }
    }
}
```

Keep in mind that it is advised to use multiple brokers in a production environment (as stated [here](https://docs.confluent.io/platform/current/kafka/deployment.html#multi-node-configuration)).

**Comment by David-Lio** [comment]
Thank you.

Just one more thing, the first time you run the project you get a "Replaying events" message in the terminal, but if for some reason you get an error while reading a kafka message and you restart the project it does not show again and apparently its no replaying the events.

I tried replicating this using the hello-world example and throwing a new error in the user read model

user.projection.js
`    [USER_CREATED]: async (store, event) => {


        throw new Error('test error')

        const createdAt = new Date(event.timestamp);
        await store.insert(tableName, {
            id: event.aggregateId,
            email: event.payload.email,
            name: event.payload.name,
            createdAt,
            updatedAt: createdAt
        });
    },`

Then if i remove the error and i restart the project there is no "Replaying Events" message and the users table its still empty

**Comment by wesone** [comment]
There are two things...

### First one is the replay mechanism
A replay will happen if a database table is missing or its scheme changed. This is determined by the [return value of the defineTable function](https://wesone.github.io/blackrik/#/ReadModelStoreAdapter?id=return).
So on initial startup the `Users` database table does not exist yet and
```javascript
await store.defineTable('Users', userTableScheme); 
```
will trigger a replay of the events that affect the `Users` database table (e.g. the `USER_CREATED` event). After reloading there won't be another replay as the `Users` table already exists and its scheme matches the second parameter of the `store.defineTable` function. You can manually trigger a replay by deleting the `Users` table and restart the application. This would resolve your problem but of course it has to resolve automatically.

### Second thing is the Kafka retry mechanism
and we do have a little problem here.
If a consumer fails to read a Kafka message (throws an error or returns a rejecting promise), Kafka uses its [retry mechanism](https://kafka.js.org/docs/retry-detailed) to call the consumer again after some time. By default Kafka does this 5 times but this can be changed with the [retry options](https://kafka.js.org/docs/1.10.0/configuration#a-name-retry-a-default-retry). So you can use something like this inside your Blackrik config:
```javascript
eventBusAdapter: {
    module: Blackrik.ADAPTERS.EVENTBUS.Kafka
    args: {
        brokers: [
            'kafka.myproject.com:9092'
        ],
        retry: {
            retries: Number.MAX_SAFE_INTEGER // probably a bit too high
        }
    }
}
``` 
If the consumer fails, Kafka will log something like
```
{"level":"ERROR","timestamp":"2021-08-22T10:08:42.924Z","logger":"kafkajs","message":"[Runner] Error when calling eachMessage","topic":"user","partition":0,"offset":"0","stack":"Error: test error\n    at USER_CREATED ...)","error":{}}
```
and call the consumer again, however the code that Kafka calls could not execute the actual event handler again and this was a problem. [This](https://github.com/wesone/blackrik/commit/aec09e9faad63029b9ac1927a78a26ce535116c4#diff-2e27b07cadbf6abf688e279ddbd84b83584d9ae9906c5e038252e9d13c2fb074) should have fixed it.

### So to sum things up
That your log shows "Replaying events" only once is normal. That your `Users` table did not update correctly is not.
I fixed it and tried it with this code :
```javascript
let throwError = true; // somewhere before module.exports
// ...
[USER_CREATED]: async (store, event) => {
    if(throwError)
    {
        throwError = false;
        throw new Error('test error');
    }
    const createdAt = new Date(event.timestamp);
    await store.insert(tableName, {
        id: event.aggregateId,
        email: event.payload.email,
        name: event.payload.name,
        createdAt,
        updatedAt: createdAt
    });
}
```
